### PR TITLE
Implement new GitHub device authentication flow

### PR DESCRIPTION
**Pull request by Keno**
Docs: https://docs.github.com/en/free-pro-team@latest/developers/apps/authorizing-oauth-apps

Screenshots:
```
Authenticating with GitHub
To continue, we need to authenticate you with GitHub. Please nagivate to
the following page in our browser and enter below code:

     https://github.com/login/device

     #############
     # XXXX-XXXX #
     #############
```
Screenshot from 2020-10-05 01-07-15
Screenshot from 2020-10-05 01-07-28
Screenshot from 2020-10-05 01-06-16

### CODE REVIEW

**Comment by giordano** [comment]
Interestingly enough, I can't reproduce locally the failures in CI :confused:

**Comment by giordano** [comment]
Apart from the one in docker https://dev.azure.com/JuliaPackaging/BinaryBuilder.jl/_build/results?buildId=6114&view=logs&jobId=54759826-d195-5ca2-1571-21806663cd36&j=54759826-d195-5ca2-1571-21806663cd36&t=9eb882d6-5a34-5e37-1506-d0b4bf16214f, I also get stuck at the non-interactive authentication

**Review by giordano — COMMENTED** [review]
Looks overall good to me, but it would be good to understand the failure https://dev.azure.com/JuliaPackaging/BinaryBuilder.jl/_build/results?buildId=6114&view=logs&j=aaebf0e6-d7cd-53db-6fd2-ebe168039a5a&t=231c1df8-60d4-55f6-9e92-9b8d8520ea6c&l=1351
```
Wizard - Downloading: Test Failed at /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:148
  Expression: getfield.(state.source_files, :hash) == [libfoo_tarball_hash]
   Evaluated: ["c925803a9d565fc743f460ce88f3fe895571faae8dab0f0fadaa375f8986f807"] == ["980136a52971f482403168b955b71597dd75d21584c66e4a0aa258275f4889b2"]
Stacktrace:
 [1] top-level scope
   @ /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:148
 [2] top-level scope
   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1113
 [3] top-level scope
   @ /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:128
Wizard - Downloading: Test Failed at /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:174
  Expression: getfield.(state.source_files, :hash) == [libfoo_tarball_hash, libfoo_tarball_hash]
   Evaluated: ["c925803a9d565fc743f460ce88f3fe895571faae8dab0f0fadaa375f8986f807", "c925803a9d565fc743f460ce88f3fe895571faae8dab0f0fadaa375f8986f807"] == ["980136a52971f482403168b955b71597dd75d21584c66e4a0aa258275f4889b2", "980136a52971f482403168b955b71597dd75d21584c66e4a0aa258275f4889b2"]
Stacktrace:
 [1] top-level scope
   @ /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:174
 [2] top-level scope
   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Test/src/Test.jl:1113
 [3] top-level scope
   @ /data/staticfloat/bb_azp/rootfs/agent_3/_work/1/s/test/wizard.jl:128
Test Summary:        | Pass  Fail  Total
Wizard - Downloading |    9     2     11
```

I'd expect to be unrelated to this PR, but started to appear here and I can't reproduce on my system

**Comment by giordano** [comment]
I'm starting to suspect the the hash failures might be related to https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/928#issuecomment-704586943 :thinking:

File: src/wizard/github.jl [hunk 1]
```diff
@@ -29,75 +29,83 @@ end
...
+        GitHub.handle_response_error(resp)
+    end
+    reply = JSON.parse(HTTP.payload(resp, String))
+
     println(outs, """
-    To continue, we need to create a GitHub access token, so that we can do
-    things like fork Yggdrasil and create pull requests against it.  To create
-    an access token, you will be asked to enter your GitHub credentials:
-    """)
+    To continue, we need to authenticate you with GitHub. Please nagivate to
```
Comment (fredrikekre) [inline src/wizard/github.jl:49]:
```suggestion
    To continue, we need to authenticate you with GitHub. Please navigate to
```

File: src/wizard/github.jl [hunk 2]
```diff
@@ -29,75 +29,83 @@ end
...
+    the following page in our browser and enter below code:
```
Comment (fredrikekre) [inline src/wizard/github.jl:50]:
```suggestion
    the following page in your browser and enter the code below:
```

File: src/wizard/github.jl [hunk 3]
```diff
@@ -29,75 +29,83 @@ end
...
+        token = token_reply["access_token"]
 
         print(outs, strip("""
         Successfully obtained GitHub authorization token!
-        This token will be used for the rest of this BB session, however if you
-        wish to use this token permanently, you may set it in your environment
-        as via the following line in your """))
+        This token will be used for the rest of this BB session.
+        You will have to re-authenticate for any future session.
+        However, if you wish to bypass this step, you may crate a
```
Comment (fredrikekre) [inline src/wizard/github.jl:100]:
```suggestion
        However, if you wish to bypass this step, you may create a
```

**Review by staticfloat — APPROVED** [review]
When this goes green, let's merge

**Pull request merged**